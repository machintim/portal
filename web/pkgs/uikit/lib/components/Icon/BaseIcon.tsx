import { memo, SVGAttributes, SVGProps } from 'react';
import { css, cx } from '@emotion/css';

export const NAME = 'ucl-icon';

interface ITitledSvgProps<T> extends SVGProps<T> {
  title?: string;
}

export interface BaseIconProps extends SVGAttributes<SVGElement> {
  svg: React.FC<ITitledSvgProps<SVGSVGElement>>;
  size?: number;
  title?: string;
  testId?: string;
}

const BaseIcon = ({
  svg,
  size = 24,
  title,
  width,
  height,
  testId = NAME,
  className,
  ...props
}: BaseIconProps) => {
  const baseStyle = css`
    vertical-align: middle;
  `;

  const style = cx(NAME, baseStyle, className);

  const Svg = svg;

  return (
    <Svg
      width={width || size}
      height={height || size}
      title={title}
      data-testid={testId}
      className={style}
      {...props}
    />
  );
};

export default memo(BaseIcon);
