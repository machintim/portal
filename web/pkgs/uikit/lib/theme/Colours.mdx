import { Meta, Title, Subtitle, ColorPalette, ColorItem } from '@storybook/addon-docs/blocks';
import theme from './defaultTheme';

<Meta title="theme/Colours" />
<Title>Colours explorer</Title>
<Subtitle>Colours approved for use in the UCL Design System</Subtitle> 

More information on the use of colour in the UCL Design System <a href="https://www.figma.com/design/8Sm5PxWOWJYpYXAzhRzUzt/UCL-Design-System-UI-Kit?node-id=1-287" target="_blank">can be found on Figma</a>. 

The colours populating this page are imported from the <a href="https://github.com/ucl-isd/uikit-react/blob/main/lib/theme/defaultTheme.ts#L5" target="_blank">theme file</a>. 

Below are all colours per category, one category per line: 

<ColorPalette>
{
  Object.entries(theme.color).map(([key, value]) => {
    return (
      <ColorItem key={key} title={key.charAt(0).toUpperCase() + key.slice(1)} colors={value} />
    );
  })
}
</ColorPalette>

Below is a breakdown of each colour category. These are the same colours as above, but grouped by category, one colour per line: 

{Object.entries(theme.color).map(([key, value]) => {
  return (
    <div key={key}>
      <h2>{key.charAt(0).toUpperCase() + key.slice(1)}</h2>

      <ColorPalette>

      {Object.entries(value).map(([colorKey, colorValue]) => {
        return (
          <ColorItem key={colorKey} title={colorKey} colors={{ [colorKey]: colorValue }} />
        );
      })}

      {console.log(key, value)}

      </ColorPalette>
    </div>
  );
})}
