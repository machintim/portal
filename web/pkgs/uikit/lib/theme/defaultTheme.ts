// UCL Theme

// base colours

const purple200 = '#2C0442'; // UCL dark purple
const purple50 = '#500778'; // UCL vibrant purple
const purple40 = '#6345A5';
const purple20 = '#BBAACC';
// TODO: where is UCL muted purple = #C6B0BC ?
const green100 = '#006B0F';
const green70 = '#008613';
const green40 = '#80C289';
const green10 = '#E6F3E8';
const green200 = '#113B3A'; // UCL dark green
const green50 = '#52C152'; // UCL vibrant green
const green20 = '#C9D1A8'; // UCL muted green

const blue100 = '#004089';
const blue80 = '#004EA8';
const blue70 = '#0D68CF';
const blue40 = '#7FB7F7';
const blue10 = '#E5F0FD';
const blue5 = '#F2F8FE';
const blue200 = '#002248'; // UCL dark blue
const blue50 = '#34C6C6'; // UCL vibrant blue
const blue20 = '#B6DCE5'; // UCL muted blue

const yellow200 = '#504D48'; // UCL dark yellow
const yellow50 = '#FFCA36'; // UCL vibrant yellow
const yellow20 = '#DAD6CA'; // UCL muted yellow

const orange100 = '#CC7100';
const orange70 = '#FF8D00';
const orange40 = '#FFC680';
const orange10 = '#FFF4E6';

const red100 = '#811712';
const red70 = '#CA0007';
const red50 = '#E48083';
const red40 = '#E48083';
const red10 = '#FAE6E7';

const pink200 = '#4B0A32'; // UCL dark pink
const pink50 = '#AC145A'; // UCL vibrant pink
const pink20 = '#DEB8C3'; // UCL vibrant pink

const black100 = '#000000';
const black90 = '#1A1A1A';
const black80 = '#333333';
const black60 = '#666666';
const black40 = '#999999';
const black20 = '#CCCCCC';
const black15 = '#D9D9D9';
const black10 = '#E6E6E6';
const black5 = '#F2F2F2';

const white = '#ffffff';

const baseColour = {
  purple200,
  purple50,
  purple40,
  purple20,

  green100,
  green70,
  green40,
  green10,
  green200,
  green50,
  green20,

  blue100,
  blue80,
  blue70,
  blue40,
  blue10,
  blue5,
  blue200,
  blue50,
  blue20,

  yellow200,
  yellow50,
  yellow20,

  orange100,
  orange70,
  orange40,
  orange10,
  // missing orange50?
  red100,
  red70,
  red50,
  red40,
  red10,

  pink200,
  pink50,
  pink20,

  black100,
  black90,
  black80,
  black60,
  black40,
  black20,
  black15,
  black10,
  black5,

  white,
};

const textColours = {
  primary: baseColour.black90, // Primary text, such as body copy, headers, displays, labels.
  secondary: baseColour.black60, // Secondary text, such as stand-first, paragraph summary, breadcrumbs.
  muted: baseColour.black40, // Tertiary text, such as input fields placeholder.
  inverted: baseColour.white, // Text on bold background, for example primary buttons label.

  interactive: baseColour.blue70, // Link, secondary button.
  default: baseColour.black60, // Status default for component label: tabs, toolbar.
  disabledOnBg: baseColour.black40, // Status disabled on background.
  disabled: baseColour.black20, // Status disabled.

  active: baseColour.blue70, // Status active for components' label: tabs, toolbar.
  hover: baseColour.blue100, // Status hover for links and components' label.
  visited: baseColour.purple40, // Status visited for links.

  error: baseColour.red70, // Label for buttons or link with destructive purpose.
  errorHover: baseColour.red100, // Error hover.
  errorDisabled: baseColour.red50, // Error disabled.
};

const interactionColours = {
  blue100: baseColour.blue100,
  blue80: baseColour.blue80,
  blue70: baseColour.blue70,
  blue40: baseColour.blue40,
  blue10: baseColour.blue10,
  blue5: baseColour.blue5,
};

const displayColours = {
  purpleDark: baseColour.purple200,
  purpleVibrant: baseColour.purple50,
  purpleMuted: baseColour.purple20,

  greenDark: baseColour.green200,
  greenVibrant: baseColour.green50,
  greenMuted: baseColour.green20,

  yellowDark: baseColour.yellow200,
  yellowVibrant: baseColour.yellow50,
  yellowMuted: baseColour.yellow20,

  pinkDark: baseColour.pink200,
  pinkVibrant: baseColour.pink50,
  pinkMuted: baseColour.pink20,

  blueDark: baseColour.blue200,
  blueVibrant: baseColour.blue50,
  blueMuted: baseColour.blue20,
};

const systemColours = {
  red100: baseColour.red100,
  red70: baseColour.red70,
  red40: baseColour.red40,
  red10: baseColour.red10,

  orange100: baseColour.orange100,
  orange70: baseColour.orange70,
  orange40: baseColour.orange40,
  orange10: baseColour.orange10,

  green100: baseColour.green100,
  green70: baseColour.green70,
  green40: baseColour.green40,
  green10: baseColour.green10,

  blue100: baseColour.blue100,
  blue80: baseColour.blue80,
  blue70: baseColour.blue70,
  blue40: baseColour.blue40,
  blue10: baseColour.blue10,
  blue5: baseColour.blue5,
};

const linkColours = {
  default: baseColour.blue70,
  hover: baseColour.blue100,
  visited: baseColour.purple40,
  disabled: baseColour.black20,
};

const blackAndWhite = {
  black: baseColour.black100,
  white: baseColour.white,
};

const neutralColours = {
  black: blackAndWhite.black,
  grey90: baseColour.black90,
  grey80: baseColour.black80,
  grey60: baseColour.black60,
  grey40: baseColour.black40,
  grey20: baseColour.black20,
  grey15: baseColour.black15,
  grey10: baseColour.black10,
  grey5: baseColour.black5,
  white: blackAndWhite.white,
};

const overlayColours = {
  blanket: 'rgba(0, 0, 0, 0.5)',
};

const color = {
  text: textColours,
  interaction: interactionColours,
  display: displayColours,
  system: systemColours,
  link: linkColours,
  neutral: neutralColours,
  overlay: overlayColours,
};

// These colour definitions will need to be refactored to match the design system.

const baseSizes = {
  s1: '1px',
  s2: '2px',
  s4: '4px',
  s6: '6px',
  s8: '8px',
  s12: '12px',
  s16: '16px',
  s20: '20px',
  s24: '24px',
  s28: '28px',
  s32: '32px',
  s36: '36px',
  s40: '40px',
  s44: '44px',
  s48: '48px',
  s52: '52px',
  s56: '56px',
  s60: '60px',
  s64: '64px',
  s80: '80px',
};

const margin = {
  m2: baseSizes.s2,
  m4: baseSizes.s4,
  m6: baseSizes.s6,
  m8: baseSizes.s8,
  m12: baseSizes.s12,
  m16: baseSizes.s16,
  m20: baseSizes.s20,
  m24: baseSizes.s24,
  m32: baseSizes.s32,
  m40: baseSizes.s40,
  m48: baseSizes.s48,
  m56: baseSizes.s56,
  m64: baseSizes.s64,
};

const padding = {
  p2: baseSizes.s2,
  p4: baseSizes.s4,
  p6: baseSizes.s6,
  p8: baseSizes.s8,
  p12: baseSizes.s12,
  p16: baseSizes.s16,
  p20: baseSizes.s20,
  p24: baseSizes.s24,
  p32: baseSizes.s32,
  p40: baseSizes.s40,
  p48: baseSizes.s48,
  p56: baseSizes.s56,
  p64: baseSizes.s64,
};

const breakpoints = {
  mobile: 320,
  tablet: 672,
  desktop: 1056,
  desktopLarge: 1536,
};

const createMediaQuery = (breakpoint: number) =>
  // Breakpoints apply at the specified width and above
  `@media (min-width: ${breakpoint}px)`;

const custom = (breakpoint: number) =>
  // Fallback for edge cases
  createMediaQuery(breakpoint);

const mq = {
  mobile: createMediaQuery(breakpoints.mobile),
  tablet: createMediaQuery(breakpoints.tablet),
  desktop: createMediaQuery(breakpoints.desktop),
  desktopLarge: createMediaQuery(breakpoints.desktopLarge),
  custom,
};

const theme = {
  color,

  thickness: {
    t1: baseSizes.s1,
    t2: baseSizes.s2,
  },

  width: {
    h16: baseSizes.s16,
    h20: baseSizes.s20,
    h24: baseSizes.s24,
  },

  height: {
    h16: baseSizes.s16,
    h20: baseSizes.s20,
    h24: baseSizes.s24,
    h32: baseSizes.s32,
    h36: baseSizes.s36,
    h40: baseSizes.s40,
    h80: baseSizes.s80,
  },

  margin,

  padding,

  radius: {
    r2: baseSizes.s2,
    r4: baseSizes.s4,
    r8: baseSizes.s8,
    r12: baseSizes.s12,
    r16: baseSizes.s16,
  },

  border: {
    b1: baseSizes.s1,
    b2: baseSizes.s2,
  },

  outline: {
    o2: baseSizes.s2,
    o4: baseSizes.s4,
  },

  font: {
    family: {
      primary: 'Helvetica, "Helvetica Neue", Arial, sans-serif',
      secondary:
        'Garamond, Baskerville, "Baskerville Old Face", "Hoefler Text", "Times New Roman", serif',
      fixed: 'Courier, fixed',
    },
    size: {
      f12: '12px',
      f14: '14px', // Font should be at least 14px in most cases
      f16: '16px', // Prefer 16px to 14px in most cases
      f18: '18px',
      f20: '20px',
      f22: '22px',
      f24: '24px',
      f26: '26px',
      f28: '28px',
      f30: '30px',
      f32: '32px',
      f36: '36px',
      f40: '40px',
    },
    weight: {
      light: 300,
      regular: 400,
      medium: 500,
      semiBold: 600,
      bold: 700,
      extraBold: 800,
    },
    lineHeight: {
      h100: '100%',
      h110: '110%',
      h120: '120%',
      h130: '130%',
      h140: '140%',
      h150: '150%',
    },
  },

  boxShadow: {
    // offset-x | offset-y | blur-radius | spread-radius | color
    y1: '0 1px 4px 1px rgba(0, 0, 0, 0.1)',
    y2: '0 2px 4px 1px rgba(0, 0, 0, 0.1)',
    x2y4: '2px 4px 8px 4px rgba(0, 0, 0, 0.1)',
    focus: `0px 0px 0px 2px ${color.neutral.white}, 0px 0px 0px 4px ${color.neutral.black}, 0px 0px 0px 6px ${color.system.orange70}`,
    active: `0px 6px 0px 0px ${color.interaction.blue80} inset`,
    destructiveActive: `0px 6px 0px 0px ${color.system.red100} inset`,
  },

  breakpoints,
  mq,

  // For ease of use, margin and padding are added at the top level
  // Usage: margin-top: ${theme.m16};
  ...margin,
  ...padding,
};

export type ThemeType = typeof theme;
export type DisplayColoursType = typeof displayColours;
export type BlackAndWhiteColoursType = typeof blackAndWhite;

export default theme;
