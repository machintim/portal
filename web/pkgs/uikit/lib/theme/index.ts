export { default as theme } from './defaultTheme';
export type {
  ThemeType,
  DisplayColoursType,
  BlackAndWhiteColoursType,
} from './defaultTheme';
export {
  default as useTheme,
  ThemeContextProvider,
} from './useTheme';
