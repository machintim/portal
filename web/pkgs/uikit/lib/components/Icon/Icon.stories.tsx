import type { Meta, StoryObj } from '@storybook/react';
import Icon from './Icon';
import { ComponentType } from 'react';

const meta: Meta = {
  title: 'Components/Work in progress/Icon',
  component: Icon,
  parameters: {
    // Appearing automatically, due to composition
    controls: { exclude: ['svg'] },
  },
  argTypes: {
    icon: {
      control: 'select',
      options: [
        'AlertTriangle',
        'ArrowLeft',
        'Avatar',
        'Calendar',
        'CheckCircle',
        'Check',
        'CheckCircle',
        'ChevronDown',
        'ChevronLeft',
        'ChevronRight',
        'ChevronUp',
        'Code',
        'Database',
        'DownloadCloud',
        'Edit',
        'Grid',
        'HardDrive',
        'HelpCircle',
        'Home',
        'Image',
        'Info',
        'Layout',
        'List',
        'Lock',
        'LogOut',
        'MapPin',
        'Menu',
        'MousePointer',
        'Plus',
        'Printer',
        'Search',
        'Settings',
        'Share2',
        'Shield',
        'Terminal',
        'Upload',
        'User',
        'XCircle',
        'XSquare',
        'X',
      ],
    },
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  name: '(Pick an icon to view)',
  args: {
    icon: 'MapPin',
    color: 'currentColor',
  },
  render: (args) => {
    const IconComponent = Icon[args.icon as keyof typeof Icon] as ComponentType;
    return (
      <>
        <IconComponent {...args} />
      </>
    );
  },
};
