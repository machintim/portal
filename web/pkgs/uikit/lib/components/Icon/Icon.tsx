import { memo } from 'react';

import AlertTriangleSvg from './svgs/AlertTriangleSvg';
import ArrowLeftSvg from './svgs/ArrowLeftSvg';
import AvatarSvg from './svgs/AvatarSvg';
import CalendarSvg from './svgs/CalendarSvg';
import CheckCircleSvg from './svgs/CheckCircleSvg';
import CheckSvg from './svgs/CheckSvg';
import ChevronDownSvg from './svgs/ChevronDownSvg';
import ChevronLeftSvg from './svgs/ChevronLeftSvg';
import ChevronRightSvg from './svgs/ChevronRightSvg';
import ChevronUpSvg from './svgs/ChevronUpSvg';
import CodeSvg from './svgs/CodeSvg';
import DatabaseSvg from './svgs/DatabaseSvg';
import DownloadCloudSvg from './svgs/DownloadCloudSvg';
import Edit2Svg from './svgs/Edit2Svg';
import GridSvg from './svgs/GridSvg';
import HardDriveSvg from './svgs/HardDriveSvg';
import HelpCircleSvg from './svgs/HelpCircleSvg';
import HomeSvg from './svgs/HomeSvg';
import ImageSvg from './svgs/ImageSvg';
import InfoSvg from './svgs/InfoSvg';
import LayoutSvg from './svgs/LayoutSvg';
import ListSvg from './svgs/ListSvg';
import LockSvg from './svgs/LockSvg';
import LogOutSvg from './svgs/LogOutSvg';
import MapPinSvg from './svgs/MapPinSvg';
import Maximize2Svg from './svgs/Maximize2Svg';
import MenuSvg from './svgs/MenuSvg';
import MessageSquareSvg from './svgs/MessageSquareSvg';
import MinusSvg from './svgs/MinusSvg';
import MousePointerSvg from './svgs/MousePointerSvg';
import PlusSvg from './svgs/PlusSvg';
import PrinterSvg from './svgs/PrinterSvg';
import SearchSvg from './svgs/SearchSvg';
import SettingsSvg from './svgs/SettingsSvg';
import Share2Svg from './svgs/Share2Svg';
import ShieldSvg from './svgs/ShieldSvg';
import TerminalSvg from './svgs/TerminalSvg';
import UploadSvg from './svgs/UploadSvg';
import UserSvg from './svgs/UserSvg';
import XCircleSvg from './svgs/XCircleSvg';
import XSquareSvg from './svgs/XSquareSvg';
import XSvg from './svgs/XSvg';
// add imports SVGs here

// social
import FacebookSvg from './svgs/social/FacebookSvg';
import FlickrSvg from './svgs/social/FlickrSvg';
import InstagramSvg from './svgs/social/InstagramSvg';
import SoundcloudSvg from './svgs/social/SoundcloudSvg';
import TiktokSvg from './svgs/social/TiktokSvg';
import XSocialSvg from './svgs/social/XSvg';
import YoutubeSvg from './svgs/social/YoutubeSvg';

// internal
import EmptySvg from './svgs/internal/EmptySvg';

import BaseIcon, { BaseIconProps } from './BaseIcon';

export interface IconProps extends BaseIconProps {}

const Icon = ({ ...props }: IconProps) => {
  return <BaseIcon {...props} />;
};

export interface SpecificIconProps extends Omit<BaseIconProps, 'svg'> {}

const AlertTriangleIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={AlertTriangleSvg}
    {...props}
  />
);
const ArrowLeftIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={ArrowLeftSvg}
    {...props}
  />
);
const AvatarIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={AvatarSvg}
    {...props}
  />
);
const CalendarIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={CalendarSvg}
    {...props}
  />
);
const CheckCircleIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={CheckCircleSvg}
    {...props}
  />
);
const CheckIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={CheckSvg}
    {...props}
  />
);
const ChevronDownIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={ChevronDownSvg}
    {...props}
  />
);
const ChevronLeftIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={ChevronLeftSvg}
    {...props}
  />
);
const ChevronRightIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={ChevronRightSvg}
    {...props}
  />
);
const ChevronUpIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={ChevronUpSvg}
    {...props}
  />
);
const CodeIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={CodeSvg}
    {...props}
  />
);
const DatabaseIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={DatabaseSvg}
    {...props}
  />
);
const DownloadCloudIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={DownloadCloudSvg}
    {...props}
  />
);
const Edit2Icon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={Edit2Svg}
    {...props}
  />
);
const GridIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={GridSvg}
    {...props}
  />
);
const HardDriveIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={HardDriveSvg}
    {...props}
  />
);
const HelpCircleIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={HelpCircleSvg}
    {...props}
  />
);
const HomeIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={HomeSvg}
    {...props}
  />
);
const ImageIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={ImageSvg}
    {...props}
  />
);
const InfoIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={InfoSvg}
    {...props}
  />
);
const LayoutIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={LayoutSvg}
    {...props}
  />
);
const ListIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={ListSvg}
    {...props}
  />
);
const LockIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={LockSvg}
    {...props}
  />
);
const LogOutIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={LogOutSvg}
    {...props}
  />
);
const MapPinIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={MapPinSvg}
    {...props}
  />
);
const Maximize2Icon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={Maximize2Svg}
    {...props}
  />
);
const MenuIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={MenuSvg}
    {...props}
  />
);
const MessageSquareIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={MessageSquareSvg}
    {...props}
  />
);
const MinusIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={MinusSvg}
    {...props}
  />
);
const MousePointerIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={MousePointerSvg}
    {...props}
  />
);
const PlusIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={PlusSvg}
    {...props}
  />
);
const PrinterIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={PrinterSvg}
    {...props}
  />
);
const SearchIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={SearchSvg}
    {...props}
  />
);
const SettingsIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={SettingsSvg}
    {...props}
  />
);
const Share2Icon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={Share2Svg}
    {...props}
  />
);
const ShieldIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={ShieldSvg}
    {...props}
  />
);
const TerminalIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={TerminalSvg}
    {...props}
  />
);
const UploadIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={UploadSvg}
    {...props}
  />
);
const UserIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={UserSvg}
    {...props}
  />
);
const XCircleIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={XCircleSvg}
    {...props}
  />
);
const XSquareIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={XSquareSvg}
    {...props}
  />
);
const XIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={XSvg}
    {...props}
  />
);
// add more icons here

// social
const FacebookIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={FacebookSvg}
    {...props}
  />
);
const FlickrIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={FlickrSvg}
    {...props}
  />
);
const InstagramIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={InstagramSvg}
    {...props}
  />
);
const SoundcloudIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={SoundcloudSvg}
    {...props}
  />
);
const TiktokIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={TiktokSvg}
    {...props}
  />
);
const XSocialIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={XSocialSvg}
    {...props}
  />
);
const YoutubeIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={YoutubeSvg}
    {...props}
  />
);

// internal
const EmptyIcon = ({ ...props }: SpecificIconProps) => (
  <Icon
    svg={EmptySvg}
    {...props}
  />
);

const MemoIcon = memo(Icon);

const specificIcons = {
  AlertTriangle: AlertTriangleIcon,
  ArrowLeft: ArrowLeftIcon,
  Avatar: AvatarIcon,
  Calendar: CalendarIcon,
  Check: CheckIcon,
  CheckCircle: CheckCircleIcon,
  ChevronDown: ChevronDownIcon,
  ChevronLeft: ChevronLeftIcon,
  ChevronRight: ChevronRightIcon,
  ChevronUp: ChevronUpIcon,
  Code: CodeIcon,
  Database: DatabaseIcon,
  DownloadCloud: DownloadCloudIcon,
  Edit: Edit2Icon,
  Grid: GridIcon,
  HardDrive: HardDriveIcon,
  HelpCircle: HelpCircleIcon,
  Home: HomeIcon,
  Image: ImageIcon,
  Info: InfoIcon,
  Layout: LayoutIcon,
  List: ListIcon,
  Lock: LockIcon,
  LogOut: LogOutIcon,
  MapPin: MapPinIcon,
  Maximize2: Maximize2Icon,
  Menu: MenuIcon,
  MessageSquare: MessageSquareIcon,
  Minus: MinusIcon,
  MousePointer: MousePointerIcon,
  Plus: PlusIcon,
  Printer: PrinterIcon,
  Search: SearchIcon,
  Settings: SettingsIcon,
  Share2: Share2Icon,
  Shield: ShieldIcon,
  Terminal: TerminalIcon,
  Upload: UploadIcon,
  User: UserIcon,
  XCircle: XCircleIcon, // << Typo here. Should be XCircle (note the missing 'c')
  XSquare: XSquareIcon,
  X: XIcon,
  // expose icons here

  // social
  Facebook: FacebookIcon,
  Flickr: FlickrIcon,
  Instagram: InstagramIcon,
  Soundcloud: SoundcloudIcon,
  Tiktok: TiktokIcon,
  XSocial: XSocialIcon,
  Youtube: YoutubeIcon,

  // Internal
  Empty: EmptyIcon,
};

export type SpecificIconsType = typeof specificIcons;

const ExtendedIconConfig = MemoIcon as typeof MemoIcon & SpecificIconsType;

let iconKey: keyof SpecificIconsType;
for (iconKey in specificIcons) {
  ExtendedIconConfig[iconKey] = specificIcons[iconKey];
}

export default ExtendedIconConfig;
