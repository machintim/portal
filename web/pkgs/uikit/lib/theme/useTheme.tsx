// We will probably move away from context/provider to store/hook for accessing theme.

import {
  createContext,
  useState,
  useContext,
  ReactNode,
} from 'react';

import theme, { ThemeType } from './defaultTheme';

const useThemeState = () => useState(theme as ThemeType);

const ThemeContext = createContext<ReturnType<
  typeof useThemeState
> | null>(null);

export const ThemeContextProvider: React.FC<{
  children: ReactNode;
}> = ({ children }) => (
  <ThemeContext.Provider value={useThemeState()}>
    {children}
  </ThemeContext.Provider>
);

const useTheme = () => {
  const value = useContext(ThemeContext);
  if (value === null) {
    throw new Error('Please add ThemeContextProvider');
  }
  return value;
};

export default useTheme;
