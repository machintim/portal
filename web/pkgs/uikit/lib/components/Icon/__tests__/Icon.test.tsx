import { describe, expect, test } from 'vitest';
import { render } from '@testing-library/react';
import Icon from '../Icon';
import { ThemeContextProvider } from '../../../theme/useTheme';

describe('Icon', () => {
  // Snapshot tests

  test('snapshot: no props', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <Icon.Check />
      </ThemeContextProvider>
    );
    expect(
      renderResult.container.firstChild
    ).toMatchSnapshot();
  });

  test('snapshot: size prop', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <Icon.Check size={32} />
      </ThemeContextProvider>
    );
    expect(
      renderResult.container.firstChild
    ).toMatchSnapshot();
  });

  test('snapshot: className prop', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <Icon.Check className='a-class-name' />
      </ThemeContextProvider>
    );
    expect(
      renderResult.container.firstChild
    ).toMatchSnapshot();
  });
});
