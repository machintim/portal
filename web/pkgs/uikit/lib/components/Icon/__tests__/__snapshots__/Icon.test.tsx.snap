// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Icon > snapshot: className prop 1`] = `
<svg
  class="ucl-icon css-148hpxb a-class-name"
  data-testid="ucl-icon"
  fill="none"
  height="24"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="3"
  viewBox="0 0 24 24"
  width="24"
  xmlns="http://www.w3.org/2000/svg"
>
  <polyline
    points="20 6 9 17 4 12"
  />
</svg>
`;

exports[`Icon > snapshot: no props 1`] = `
<svg
  class="ucl-icon css-148hpxb"
  data-testid="ucl-icon"
  fill="none"
  height="24"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="3"
  viewBox="0 0 24 24"
  width="24"
  xmlns="http://www.w3.org/2000/svg"
>
  <polyline
    points="20 6 9 17 4 12"
  />
</svg>
`;

exports[`Icon > snapshot: size prop 1`] = `
<svg
  class="ucl-icon css-148hpxb"
  data-testid="ucl-icon"
  fill="none"
  height="32"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="3"
  viewBox="0 0 24 24"
  width="32"
  xmlns="http://www.w3.org/2000/svg"
>
  <polyline
    points="20 6 9 17 4 12"
  />
</svg>
`;
